File: proj/v8seg/opencv/main.py
import cv2
import numpy as np
import random

def sigmoid(x):
    return 1 / (1 + np.exp(-x))


class YOLOV8SEG:

    def __init__(self, modelpath, conf_thres=0.7, nms_thres=0.8):
        self.confThreshold = conf_thres
        self.nmsThreshold = nms_thres

        self.class_names = [x.strip() for x in open("coco.names", "r").readlines()] 
        self.net = cv2.dnn.readNet(modelpath)
        self.inpHeight = 640
        self.inpWidth = 640
        
        self.nm = 32

    def drawPred(self, image, box, clsid, conf):
        x,y,w,h = box

        cv2.rectangle(image, (x,y), (x+w,y+h), (144,125,0), 2)

        label = self.class_names[clsid] + ":" + str(round(conf, 2))

        cv2.putText(image, label, (x,y-10), cv2.FONT_HERSHEY_SIMPLEX, 1, (90,255,67), 2)

    def detect(self, image):

        blob = cv2.dnn.blobFromImage(image, 1.0 / 255.0, (self.inpHeight, self.inpWidth), (0,0,0), True, False)

        self.net.setInput(blob)

        outs = self.net.forward(self.net.getUnconnectedOutLayersNames())

        ratiow, ratioh = image.shape[1] / self.inpWidth, image.shape[0] / self.inpHeight
        
        boxes, confs, ids, m_preds = [], [], [], []
        for pred in np.squeeze(outs[0]).transpose():
            maxss = np.max(pred[4:-self.nm])
            idp = np.argmax(pred[4:-self.nm])
            if maxss >= self.confThreshold:
                cx = pred[0] * ratiow
                cy = pred[1] * ratioh
                w = pred[2] * ratiow
                h = pred[3] * ratioh

                left = int(cx - 0.5 * w)
                top = int(cy - 0.5 * h)

                boxes.append([left, top, int(w), int(h)])
                confs.append(maxss)
                ids.append(idp)
                m_preds.append(list(pred[-self.nm:]))

        indices = cv2.dnn.NMSBoxes(boxes, confs, self.confThreshold, self.nmsThreshold)

        np_boxes = np.array(boxes)[indices]
        mask_preds = np.array(m_preds)[indices]
        mask_output = np.squeeze(outs[1])

        image = self.maskimg2pred(mask_preds, np_boxes, mask_output, image)

        for idx in indices:
            self.drawPred(image, boxes[idx], ids[idx], confs[idx])
        return image

    def maskimg2pred(self, mask_preds, np_boxes, mask_out, image):

        nm_mask, mask_height, mask_width = mask_out.shape

        masks = sigmoid(mask_preds @ mask_out.reshape((nm_mask, -1))).reshape((-1, mask_height, mask_width))

        mbratioh, mbratiow = image.shape[0] / mask_height, image.shape[1] / mask_width

        mask_img = image.copy()
        scale_boxes = np_boxes.copy().astype(np.float32)

        scale_boxes[:, [0,2]] /= mbratiow
        scale_boxes[:, [1,3]] /= mbratioh

        blur_size = (int(mbratiow), int(mbratioh))

        for i in range(len(np_boxes)):
            sx, sy, sw, sh = scale_boxes[i].astype(int)
            ox, oy, ow, oh = np_boxes[i].astype(int)

            scale_crop_mask = masks[i][sy:(sy+sh), sx:(sx+sw)]
            crop_mask = cv2.resize(scale_crop_mask, (ow,oh), cv2.INTER_CUBIC)

            crop_mask = cv2.blur(crop_mask, blur_size)
            crop_mask = (crop_mask > 0.5).astype(np.uint8)[..., np.newaxis]

            crop_mask_img = mask_img[oy:(oy+oh), ox:(ox+ow)]

            mask_img[oy:(oy+oh), ox:(ox+ow)] = crop_mask_img * (1 - crop_mask) + tuple(random.sample(range(255), 3)) * crop_mask
        mask_alpha = 0.5
        return cv2.addWeighted(mask_img, mask_alpha, image, 1-mask_alpha, 0)


if __name__ == "__main__":


    net = YOLOV8SEG("weights/yolov8n-seg.onnx")

    srcimg = cv2.imread("imgs/person.jpg")

    tarimg = net.detect(srcimg)

    cv2.imwrite("imgs/person_opencv_py.jpg", tarimg)

File: proj/v8seg/opencv/main.cpp
#include <iostream>
#include <fstream>
#include <opencv2/imgproc.hpp>
#include <opencv2/dnn.hpp>
#include <opencv2/highgui.hpp>

class YOLOV8SEG
{
public:
	YOLOV8SEG(float confThreshold, float nmsThreshold, std::string modelpath);
	void drawPred(cv::Mat& frame, cv::Rect box, int classid, float conf);
	void maskimg2Pred(cv::Mat& frame, cv::Mat mask_preds, std::vector<cv::Rect> np_boxes, cv::Mat maskout);
	void detect(cv::Mat& frame);
private:
	int inpWidth;
	int inpHeight;

	int nm;
	float confThreshold;
	float nmsThreshold;

	cv::dnn::Net net;

	std::vector<std::string> class_names;
};



YOLOV8SEG::YOLOV8SEG(float confThreshold, float nmsThreshold, std::string modelpath)
{
	this->confThreshold = confThreshold;
	this->nmsThreshold = nmsThreshold;

	this->inpWidth = 640;
	this->inpHeight = 640;
	this->nm = 32;

	this->net = cv::dnn::readNet(modelpath);

	std::ifstream ifs("coco.names");
	std::string line;

	while(getline(ifs, line)) this->class_names.push_back(line.substr(0, line.length()-1));

}

void YOLOV8SEG::drawPred(cv::Mat& frame, cv::Rect box, int classid, float conf)
{
	cv::rectangle(frame, box, cv::Scalar(0,255,144), 2);

	std::string label = cv::format("%.2f", conf);

	label = this->class_names[classid] + ":" + label;

	cv::putText(frame, label, cv::Point(box.x, box.y-10), cv::FONT_HERSHEY_SIMPLEX, 1, cv::Scalar(144,255,0), 2);


}

void YOLOV8SEG::maskimg2Pred(cv::Mat& frame, cv::Mat mask_preds, std::vector<cv::Rect> np_boxes, cv::Mat maskout)
{
	int num_mask = maskout.size[1], mask_height = maskout.size[2], mask_width = maskout.size[3];
	float* maskptr = (float*)maskout.data;
	cv::Mat protos = cv::Mat(num_mask, mask_height*mask_width, CV_32F, maskptr);
	cv::Mat maskRes = (mask_preds * protos).t();
	cv::Mat masks = maskRes.reshape(np_boxes.size(), {mask_height, mask_width});

	float mbratiow = (float)frame.cols / mask_width, mbratioh = (float)frame.rows / mask_height;
	cv::Size blur_size = cv::Size(int(mbratiow), int(mbratioh));
	cv::Mat mask_img = frame.clone();

	std::vector<cv::Mat> maskChannels;
	cv::split(masks, maskChannels);
	cv::RNG rng(345);

	for(int i = 0; i < np_boxes.size(); ++i)
	{
		int ox = np_boxes[i].x, oy = np_boxes[i].y, ow = np_boxes[i].width, oh = np_boxes[i].height;
		int sx = int(ox / mbratiow), sy = int(oy / mbratioh), sw = int(ow / mbratiow), sh = int(oh / mbratioh);

		cv::Mat dest;
		cv::exp(-maskChannels[i], dest);
		dest = 1 / (1 + dest);

		cv::Rect roi(sx, sy, sw, sh);
		cv::Mat scale_crop_mask = dest(roi);

		cv::Mat crop_mask;
		cv::resize(scale_crop_mask, crop_mask, cv::Size(ow, oh), cv::INTER_CUBIC);
		cv::blur(crop_mask, crop_mask, blur_size);

		crop_mask = crop_mask > 0.5;
		
		int b = rng.uniform(0,255), g = rng.uniform(0,255), r = rng.uniform(0, 255);
		mask_img(np_boxes[i]).setTo(cv::Scalar(b,g,r), crop_mask);
	
	}
	cv::addWeighted(mask_img, 0.5, frame, 0.5, 0, frame);

}

void YOLOV8SEG::detect(cv::Mat& frame)
{
	cv::Mat blob = cv::dnn::blobFromImage(frame, 1.0 / 255.0, cv::Size(this->inpHeight, this->inpWidth), cv::Scalar(0,0,0), true, false);

	this->net.setInput(blob);

	std::vector<cv::Mat> outs;

	this->net.forward(outs, this->net.getUnconnectedOutLayersNames());

	int numproposal = outs[0].size[2];
	int nout = outs[0].size[1];
	outs[0] = outs[0].reshape(0, nout);

	std::vector<int> ids;
	std::vector<float> confs;
	std::vector<cv::Rect> boxes;
	std::vector<cv::Mat> m_preds;

	float ratiow = (float)frame.cols / this->inpWidth;
	float ratioh = (float)frame.rows / this->inpHeight;
	cv::transpose(outs[0], outs[0]);
	float* pdata = (float*)outs[0].data;

	for(int n = 0; n < numproposal; ++n)
	{
		float maxss = 0.0;
		int idp = 0;
		for(int k = 0; k < nout - 4 - this->nm; ++k)
		{
			if(pdata[k + 4] > maxss)
			{
				maxss = pdata[k + 4];
				idp = k;
			}
		
		}
		if(maxss >= this->confThreshold)
		{
			float cx = pdata[0] * ratiow;
			float cy = pdata[1] * ratioh;
			float w = pdata[2] * ratiow;
			float h = pdata[3] * ratioh;

			int left = int(cx - 0.5 * w);
			int top = int(cy - 0.5 * h);

			boxes.push_back(cv::Rect(left, top, int(w), int(h)));
			confs.push_back(maxss);
			ids.push_back(idp);
			m_preds.push_back(outs[0].row(n).colRange(nout-this->nm, nout));
		
		}
	
		pdata += nout;
	}
	
	std::vector<int> indices;
	cv::dnn::NMSBoxes(boxes, confs, this->confThreshold, this->nmsThreshold, indices);

	std::vector<cv::Rect> np_boxes;
	cv::Mat mask_preds;

	for(int idx : indices)
	{
		m_preds[idx].convertTo(m_preds[idx], CV_32F);
		np_boxes.push_back(boxes[idx]);
		mask_preds.push_back(m_preds[idx]);
	
	}

	this->maskimg2Pred(frame, mask_preds, np_boxes, outs[1]);

	for(int idx : indices) this->drawPred(frame, boxes[idx], ids[idx], confs[idx]);
}


int main()
{
	YOLOV8SEG net(0.7, 0.8, "weights/yolov8n-seg.onnx");

	cv::Mat srcimg = cv::imread("imgs/person.jpg");

	net.detect(srcimg);

	cv::imwrite("imgs/person_opencv_cpp.jpg", srcimg);

	return 0;


}











